backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailAggregator.java

```diff
@@ -0,0 +1,123 @@
+                .onItem().transformToUniAndConcatenate(eventType -> Uni.combine().all().unis(
+                        Uni.createFrom().item(eventType),
+                        endpointResources.getTargetEndpointsFromType(
+                                aggregationKey.getAccountId(),
+                                aggregationKey.getBundle(),
+                                aggregationKey.getApplication(),
+                                eventType,
+                                EndpointType.EMAIL_SUBSCRIPTION
+                        ).onItem().transform(Set::copyOf)
+                ).asTuple());
+
+        Uni<Map<String, Set<User>>> subscribersByEventTypeUni = subscriptionResources
+                .getEmailSubscribers(aggregationKey.getAccountId(), aggregationKey.getBundle(), aggregationKey.getApplication(), emailSubscriptionType)
+                .onItem().transformToMulti(emailSubscriptions -> {
+                    Set<String> subscribers = emailSubscriptions.stream().map(EmailSubscription::getUserId).collect(Collectors.toSet());
+
+                    return endpointsPerEventTypeMulti
+                            .onItem().transformToUniAndConcatenate(
+                                endpointsPerEventType -> Uni.combine().all().unis(
+                                        Uni.createFrom().item(endpointsPerEventType.getItem1()),
+                                        recipientResolver.recipientUsers(aggregationKey.getAccountId(), endpointsPerEventType.getItem2(), subscribers)
+                                ).asTuple()
+                            );
+                }).collect().asMap(Tuple2::getItem1, Tuple2::getItem2);
+
+        return subscribersByEventTypeUni.onItem().transformToUni(subscribersByEventType ->
+                emailAggregationsUni.onItem().transform(emailAggregations -> {
+                    Map<User, AbstractEmailPayloadAggregator> aggregated = new HashMap<>();
+
+                    for (EmailAggregation emailAggregation : emailAggregations) {
+                        subscribersByEventType
+                                .get(getEventType(emailAggregation))
+                                .forEach(user -> fillUsers(aggregationKey, user, aggregated, emailAggregation));
+                    }
+
+                    return aggregated;
+                })
+        ).onItem().transform(aggregatorMap -> aggregatorMap
+                .entrySet()
+                .stream()
+                .filter(entry -> entry.getValue().getProcessedAggregations() > 0)
+                .peek(entry -> {
+                    entry.getValue().setStartTime(start);
+                    entry.getValue().setEndTimeKey(end);
+                })
+                .collect(
+                    Collectors.toMap(
+                            Map.Entry::getKey,
+                            entry -> entry.getValue().getContext()
+                    )
+                )
+        );
+    }
```

Comment:
Would this work? Sorry, I didn't mean to rewrite your code 😃 

```suggestion
    private Uni<Set<String>> getEmailSubscribers(EmailAggregationKey aggregationKey, EmailSubscriptionType emailSubscriptionType) {
        return subscriptionResources
                .getEmailSubscribers(aggregationKey.getAccountId(), aggregationKey.getBundle(), aggregationKey.getApplication(), emailSubscriptionType)
                .onItem().transform(subscriptions ->
                        // TODO Return directly the userId from the database, we don't need the EmailSubscription.
                        subscriptions.stream().map(EmailSubscription::getUserId).collect(Collectors.toSet())
                )
                // This will prevent multiple database queries executions during the aggregation process.
                .memoize().indefinitely();
    }

    public Uni<Map<User, Map<String, Object>>> getAggregated(EmailAggregationKey aggregationKey, EmailSubscriptionType emailSubscriptionType, LocalDateTime start, LocalDateTime end) {
        Map<User, AbstractEmailPayloadAggregator> aggregated = new HashMap<>();
        Uni<Set<String>> subscribers = getEmailSubscribers(aggregationKey, emailSubscriptionType);

        // First, we retrieve all aggregations that match the given key.
        return emailAggregationResources.getEmailAggregation(aggregationKey, start, end)
                .onItem().transformToMulti(Multi.createFrom()::iterable)
                // For each aggregation...
                .onItem().transformToUniAndConcatenate(aggregation -> {
                    // We need its event type to determine the target endpoints.
                    String eventType = getEventType(aggregation);
                    // Let's retrieve these targets.
                    return endpointResources.getTargetEndpointsFromType(aggregationKey.getAccountId(), aggregationKey.getBundle(), aggregationKey.getApplication(), eventType, EndpointType.EMAIL_SUBSCRIPTION)
                            .onItem().transform(Set::copyOf)
                            // Now we want to determine who will actually receive the aggregation email.
                            .onItem().transformToUni(endpoints ->
                                    // All users who subscribed to the current application and subscription type combination are recipients candidates.
                                    subscribers.onItem().transformToUni(users ->
                                            /*
                                             * The actual recipients list is a subset of the candidates.
                                             * The target endpoints properties will determine whether or not each candidate will actually receive an email.
                                             */
                                            recipientResolver.recipientUsers(aggregationKey.getAccountId(), endpoints, users)
                                    )
                            )
                            /*
                             * We now have the final recipients list.
                             * Let's populate with synchronized side-effects (invoke) the Map that will be returned by the method.
                             */
                            .onItem().invoke(users -> {
                                users.forEach(user -> {
                                    // It's aggregation time!
                                    fillUsers(aggregationKey, user, aggregated, aggregation);
                                });
                            });
                })
                // We need to wait for everything to be done before proceeding to the final step.
                .onItem().ignoreAsUni()
                // We need to return the Map that was populated using synchronized side-effects.
                .replaceWith(aggregated)
                .onItem().transform(aggregatorMap ->
                        aggregatorMap
                                .entrySet()
                                .stream()
                                .peek(entry -> {
                                    // TODO These fields could be passed to EmailPayloadAggregatorFactory.by since we know them from the beginning.
                                    entry.getValue().setStartTime(start);
                                    entry.getValue().setEndTimeKey(end);
                                })
                                .collect(
                                        Collectors.toMap(
                                                Map.Entry::getKey,
                                                entry -> entry.getValue().getContext()
                                        )
                                )
                );
    }
```

Reply:
:hocho: :page_facing_up: :drop_of_blood: :drop_of_blood: